packages/apollo-server-core/src/plugin/schemaReporting/index.ts

```diff
@@ -0,0 +1,130 @@
+import os from 'os';
+import { ApolloServerPlugin } from 'apollo-server-plugin-base';
+import { v4 as uuidv4 } from 'uuid';
+import { printSchema } from 'graphql';
+import { createHash } from 'crypto';
+import { SchemaReporter } from './schemaReporter';
+
+export interface ApolloServerPluginSchemaReportingOptions {
+  /**
+   * The schema reporter waits before starting reporting.
+   * By default, the report waits some random amount of time between 0 and 10 seconds.
+   * A longer interval leads to more staggered starts which means it is less likely
+   * multiple servers will get asked to upload the same schema.
+   *
+   * If this server runs in lambda or in other constrained environments it would be useful
+   * to decrease the schema reporting max wait time to be less than default.
+   *
+   * This number will be the max for the range in ms that the schema reporter will
+   * wait before starting to report.
+   */
+  initialDelayMaxMs?: number;
+  /**
+   * Override the reported schema that is reported to the Apollo registry.
+   * This schema does not go through any normalizations and the string is
+   * directly sent to the Apollo registry.
+   * This would be useful for comments or other ordering and whitespace
+   *  changes that get stripped when generating a `GraphQLSchema`
+   */
+  overrideReportedSchema?: string;
+  /**
+   * The URL to use for reporting schemas. Primarily for testing and internal
+   * Apollo use.
+   */
+  endpointUrl?: string;
+}
+
+export function ApolloServerPluginSchemaReporting(
+  {
+    initialDelayMaxMs,
+    overrideReportedSchema,
+    endpointUrl,
+  }: ApolloServerPluginSchemaReportingOptions = Object.create(null),
+): ApolloServerPlugin {
+  const bootId = uuidv4();
+
+  return {
+    __internal_plugin_id__() {
+      return 'SchemaReporting';
+    },
+    async serverWillStart({ apollo, schema, logger }) {
+      if (!apollo.key) {
+        throw Error(
+          'To use ApolloServerPluginSchemaReporting, you must provide an Apollo API ' +
+            'key, via the $APOLLO_KEY environment variable or via `new ApolloClient({apollo: {key})`',
+        );
+      }
+      if (!apollo.graphId) {
+        throw Error(
+          "To use ApolloServerPluginSchemaReporting, you must provide your graph's ID, " +
+            "either by using an API key starting with 'service:',  or by providing it explicitly via " +
+            'the $APOLLO_GRAPH_ID environment variable or via `new ApolloClient({apollo: {graphId}})`',
+        );
+      }
+
+      const executableSchema = overrideReportedSchema ?? printSchema(schema);
```

Comment:
Executable schema generally means an actual `GraphQLSchema`, so the naming here is a bit misleading. `printedSchema`?

Reply:
For some reason, executableSchema is the name of the parameter in our GraphQL API. So I think matching is good?